zephyr/lib/cpu.c

```diff
@@ -64,6 +68,42 @@ LOG_MODULE_DECLARE(zephyr, CONFIG_SOF_LOG_LEVEL);
 
 extern struct tr_ctx zephyr_tr;
 
+/* address where zephyr PM will save memory during D3 transition */
+#ifdef CONFIG_ADSP_IMR_CONTEXT_SAVE
+extern void *global_imr_ram_storage;
+#endif
+
+void cpu_notify_state_entry(enum pm_state state)
+{
+	if (!cpu_is_primary(arch_proc_id()))
+		return;
+
+	if (state == PM_STATE_SOFT_OFF) {
+#ifdef CONFIG_ADSP_IMR_CONTEXT_SAVE
+		size_t storage_buffer_size;
+
+		/* allocate IMR global_imr_ram_storage */
+		const struct device *tlb_dev = DEVICE_DT_GET(DT_NODELABEL(tlb));
+
+		__ASSERT_NO_MSG(tlb_dev);
+		const struct intel_adsp_tlb_api *tlb_api =
+				(struct intel_adsp_tlb_api *)tlb_dev->api;
+
+		/* get HPSRAM storage buffer size */
+		storage_buffer_size = tlb_api->get_storage_size();
+
+		/* add space for LPSRAM */
+		storage_buffer_size += LP_SRAM_SIZE;
+
+		/* allocate IMR buffer and store it in the global pointer */
+		global_imr_ram_storage = rmalloc(SOF_MEM_ZONE_SYS,
```

Comment:
By alloc.h  "* 1) System Zone. Fixed size heap where alloc always succeeds and is never freed. Used by any init code that will never give up the memory." 
This memory is freed at restore so SYS_RUNTIME would be more appropriate zone. Functionally this should not matter as SOS_MEM_CAPS_L3 is passed so actual allocation is done from IMR.